### PR TITLE
Adding '-x' option

### PR DESCRIPTION
**Pull request by Stonemason5040**
Added help for '-x' and implemented it so that the script expects faliure for test specified by argument.

### Motivation and Context

Link to feature request:
https://github.com/tomato42/tlsfuzzer/issues/430

### Checklist

- [ ] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [ ] the changes are also reflected in documentation and code comments
- [ ] all new and existing tests pass (see Travis CI results)
- [ ] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [ ] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [ ] new and modified scripts were ran against popular TLS implementations:
  - [ ] OpenSSL
  - [ ] NSS
  - [ ] GnuTLS
- [ ] required version of tlslite-ng updated in requirements.txt and README.md

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/652)

### CODE REVIEW

**Comment by Stonemason5040** [comment]
Could the `-X` work like this:
`-x 'test-name' -X 'expected-message'`
Or it should be like this:
`-X 'test-name 'expected-message'`

**Comment by Stonemason5040** [comment]
Could the `-X` work like this:
`-x 'test-name' -X 'expected-message'`
Or it should be like this:
`-X 'test-name 'expected-message'`

**Comment by tomato42** [comment]
> Could the `-X` work like this:
> `-x 'test-name' -X 'expected-message'`
> Or it should be like this:
> `-X 'test-name 'expected-message'`

I was thinking of the former approach: `-x 'test name' -X 'expected-message'`

**Comment by Stonemason5040** [comment]
could the output for Unexpected pass look something like this:
version: 4

Test end
successful: 0
expectedly_failed: 0
failed: 2
  
unexpected passes:
  'sanity'
  'sanity'

**Comment by tomato42** [comment]
I'd prefer the pytest approach: unexpected passes counted separately from failures

**Comment by Stonemason5040** [comment]
Is this the only way to get the error message and if so, what part should i use? (for the `-X` argument) `print(traceback.format_exc())`
output:
```
Traceback (most recent call last):
File "scripts/test-client-compatibility.py", line 5761, in main
runner.run()
File "/home/frank/tlsfuzzer/tlsfuzzer/runner.py", line 237, in run
RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, protocol_version)
```

**Comment by tomato42** [comment]
> @tomato42 Is this the only way to get the error message and if so, what part should i use? (for the `-X` argument) `print(traceback.format_exc())`
> output:
> 
> ```
> Traceback (most recent call last):
> File "scripts/test-client-compatibility.py", line 5761, in main
> runner.run()
> File "/home/frank/tlsfuzzer/tlsfuzzer/runner.py", line 237, in run
> RecordHeader2)))
> AssertionError: Unexpected message from peer: Alert(fatal, protocol_version)
> ```

I think just comparing against `str()` of the exception raised should be enough

**Comment by Stonemason5040** [comment]
___
*[scripts/test-client-compatibility.py, line 5794 at r7](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M2saHim-26M7tUQu-kv:-M3BsIq08BWGXx1UV-pm:bp7sk2s) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/64d17fd6e00f1526932133720e217fa144025e00/scripts/test-client-compatibility.py#L5794)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

honestly, I don't like that the check if the failure is expected or not is in 3 places with multiple negation of the result; that makes it hard to follow and reason if all cases are handled correctly

could you rework it so that it's all in a single place?
</blockquote></details>

@tomato42  Is it better this way?

**Comment by Stonemason5040** [comment]
___
*[scripts/test-client-compatibility.py, line 5791 at r11](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M3Cmeii20jTNjdV4Jge:-M3CmeijDNECtwrzBhKx:bej6vdl) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/cf7b49666149240dfb366fc78f436daaaee66baf/scripts/test-client-compatibility.py#L5791)):*
> ```Python
>                         failed.append(c_name)
>                         print("Expected error message: \'{0}\'\n"   \
>                                 .format(expected_failures[c_name]))
> ```

This displays the `-X message` just under the `traceback` so you can compare it.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-client-compatibility.py, line 5784 at r11](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M3CnJzXAUJr2SSmHMYS:-M3CnJzXAUJr2SSmHMYT:b-sc0a6f) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/cf7b49666149240dfb366fc78f436daaaee66baf/scripts/test-client-compatibility.py#L5784)):*
> ```Python
>                 xpass += 1
>                 xpassed.append(c_name)
>                 print("WRONG: expected failture but passed\n")
> ```

Also this message makes it more clear.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-client-compatibility.py, line 5791 at r10](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M3CaXmuEUezT_97uCZG:-M3CoJvf6mKndbeloANl:b-6zn33w) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/b7952b1378005b3f2bc8fcd43c6ea4ec348b9f9e/scripts/test-client-compatibility.py#L5791)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

when this `OK` is printed, it will have a backtrace above it, I wonder if printing "OK - expected failure\n" wouldn't be more user-friendly? OTOH, having the same report for both expected passes and expected failures also has some logic behind it...

Check out how it looks like in action and make a call.
</blockquote></details>

I am not sure, what do you mean by the second sentence.

**Comment by Stonemason5040** [comment]
@tomato42 Also is it better for you if i make comments under the changes i have made via **reviewable**
like i did with last changes

**Comment by Stonemason5040** [comment]
___
*[scripts/test-client-compatibility.py, line 5791 at r11](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M3Cmeii20jTNjdV4Jge:-M3GvfD99e6GO8jcPgKN:b-9uzcsn) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/cf7b49666149240dfb366fc78f436daaaee66baf/scripts/test-client-compatibility.py#L5791)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

good idea!
though I don't think either the escapes for `'` or the line break are necessary here
</blockquote></details>

I would leave the line break as it looks much better with it.

**Comment by Stonemason5040** [comment]
:lgtm:

**Comment by Stonemason5040** [comment]
___
*[tests/tlslite-ng.json, line 113 at r4](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M29Fhp47j2V4HiW2wMK:-M3L2J9j5pB1UQqMkhFn:b-mlbj82) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/3782a1ffd42cf4724ef8abac21b485f7bdaf9df8/tests/tlslite-ng.json#L113)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

because of laziness, not because it was unnecessary
</blockquote></details>

oh, ok

**Comment by Stonemason5040** [comment]
___
*[tests/tlslite-ng.json, line 120 at r4](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M29FljC9b2Vnp4SdqjO:-M3LErLi9DaY_J0oLAM_:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/3782a1ffd42cf4724ef8abac21b485f7bdaf9df8/tests/tlslite-ng.json#L120)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

same here?
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[tests/tlslite-ng.json, line 113 at r4](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M29Fhp47j2V4HiW2wMK:-M3LHASJCGHt8UYvZFIS:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/3782a1ffd42cf4724ef8abac21b485f7bdaf9df8/tests/tlslite-ng.json#L113)):*
<details><summary><i>Previously, Stonemason5040 (Frantisek Krenzelok) wrote…</i></summary><blockquote>

oh, ok
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-conversation.py, line 173 at r2](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M1RHEiP21-RALFkav2d:-M3LHtbR9odUPHgJJ8AX:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/5ab07e3a590ed623c7bc067994bd51e598744807/scripts/test-conversation.py#L173)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

we need to separately count expected failures and unexpected passes; unexpected passes should be printed in a separate list, before the `failed:` listing.
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-conversation.py, line 42 at r2](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M1RGDYI3mnEpHgNV-DS:-M3LI0SZDHHrP7Q6-MX8:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/5ab07e3a590ed623c7bc067994bd51e598744807/scripts/test-conversation.py#L42)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

Maybe something like this?:
```
    print(" -x probe-name  expect the probe to fail. When such probe passes despite being marked like this")
    print("                it will be reported in the test summary and the whole script will fail.")
    print("                May be specified multiple times.")
```
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-conversation.py, line 67 at r2](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M1QaOfq1SZMgmSjwcgr:-M3LI29W6Pm4xmfGQMHw:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/5ab07e3a590ed623c7bc067994bd51e598744807/scripts/test-conversation.py#L67)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

Please read the above and my comments in #430; if you are not sure how to implement it discuss it first, prepare a PR that changes just `test-conversation.py` second and create a PR that updates all test cases third.
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-conversation.py, line 29 at r2](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M1RHbpe1NSXVQbWzs3r:-M3LI4dl75uDS9VYvsVQ:b-41pdmq) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/5ab07e3a590ed623c7bc067994bd51e598744807/scripts/test-conversation.py#L29)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

version unupdated
</blockquote></details>

What version should i update it to, and should i update other tests as well ?

**Comment by Stonemason5040** [comment]
___
*[scripts/test-conversation.py, line 29 at r2](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M1RHbpe1NSXVQbWzs3r:-M3tyCqqEnXacwwQlbna:bcn3w4m) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/5ab07e3a590ed623c7bc067994bd51e598744807/scripts/test-conversation.py#L29)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

just increase it by one, yes, please update other tests as well
</blockquote></details>

Some test do not have `version` should i add it?

**Comment by Stonemason5040** [comment]
___
*[tests/scripts_retention.py, line 238 at r14](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M3LWPI6ERwJcbCtz1Cj:-M3uRT7y6u8BEN_xuy0P:b-896fix) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/c5dfacffa4e7c4354659185c3d1d0114b4b68f57/tests/scripts_retention.py#L238)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

now that we have actual "expected fail" in form of "xfail", I think we should skip the "expected" part in the 5 lines above
</blockquote></details>

Done.

**Comment by Stonemason5040** [comment]
___
*[scripts/test-zero-length-data.py, line 25 at r18](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M4FfyTg7FrWcTUVEn-8:-M4IwWTT3fTj-UgKqz-w:btb0ls0) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/b901805a554dfc458826a4fb01d591669116064a/scripts/test-zero-length-data.py#L25)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

huh? why that's not imported? but defined locally?
</blockquote></details>

It was like this in one of the scripts so i supposed it is like this in all of them, anyway it is fixed now

**Comment by Stonemason5040** [comment]
___
*[scripts/test-sslv2-connection.py, line 179 at r24](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M5MIR2G7NEYVuhsw9vt:-M5MIR2H7otYcnHwiptL:b-l1xfpr) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/e4bb3fc976abc766a658cf5845fe4a2aa9af3f4e/scripts/test-sslv2-connection.py#L179)):*
> ```Python
>         print("FAILED:\n\t{0}".format('\n\t'.join(repr(i) for i in sort)))
> 
>     if good > 0:
> ```

Is this intentional or is it misstake
This probably causes the `fail: 3` output of script_retention.py

**Comment by Stonemason5040** [comment]
___
*[scripts/test-sslv2-connection.py, line 121 at r24](https://reviewable.io/reviews/tomato42/tlsfuzzer/652#-M5MJHgI5oeKw6jpLvqq:-M5MJHgI5oeKw6jpLvqr:b-a29bwf) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/7f890e3b595a5e283cbbab66d04423048919b619/scripts/test-sslv2-connection.py#L121)):*
> ```Python
>     print("failed: {0}".format(bad))
> 
>     if good > 0:
> ```

Is this intentional? it might cause `fail : 3` in the output of script_retention

**Comment by Stonemason5040** [comment]
I have noticed that `xpass` doesn't render the test script as unsuccessful.
Do we want to keep this behaviour or change it?

**Comment by tomato42** [comment]
looks good, thank you!
